packages/block-library/src/heading/autogenerate-anchors.js

```diff
@@ -0,0 +1,123 @@
+/**
+ * External dependencies
+ */
+import { isNil, deburr, trim, startsWith } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { cleanForSlug } from '@wordpress/url';
+
+/**
+ * Runs a callback over all blocks, including nested blocks.
+ *
+ * @param {Object[]} blocks   The blocks.
+ * @param {Function} callback The callback.
+ *
+ * @return {void}
+ */
+const recurseOverBlocks = ( blocks, callback ) => {
+	for ( const block of blocks ) {
+		// eslint-disable-next-line callback-return
+		callback( block );
+		if ( block.innerBlocks ) {
+			recurseOverBlocks( block.innerBlocks, callback );
+		}
+	}
+};
+
+/**
+ * Returns the text without markup.
+ *
+ * @param {string} text The text.
+ *
+ * @return {string} The text without markup.
+ */
+const getTextWithoutMarkup = ( text ) => {
+	const dummyElement = document.createElement( 'div' );
+	dummyElement.innerHTML = text;
+	return dummyElement.innerText;
+};
+
+/**
+ * Get all heading anchors.
+ *
+ * @param {Object} blockList An object containing all blocks.
+ * @param {string} excludeId A block ID we want to exclude.
+ *
+ * @return {string[]} Return an array of anchors.
+ */
+export const getAllHeadingAnchors = ( blockList, excludeId ) => {
+	const anchors = [];
+
+	recurseOverBlocks( blockList, ( block ) => {
+		if (
+			block.name === 'core/heading' &&
+			( ! excludeId || block.clientId !== excludeId ) &&
+			block.attributes.anchor
+		) {
+			anchors.push( block.attributes.anchor );
+		}
+	} );
+
+	return anchors;
+};
+
+/**
+ * Generate the anchor for a heading.
+ *
+ * @param {string}   anchor            The heading anchor.
+ * @param {string}   content           The block content.
+ * @param {string[]} allHeadingAnchors An array containing all headings anchors.
+ *
+ * @return {string|null} Return the heading anchor.
+ */
+const generateAnchor = ( anchor, content, allHeadingAnchors ) => {
```

Comment:
What @youknowriad  mentioned about creating a custom hook like `useGeneratedAnchor`, I think it would be more of like this (please Riad correct me if I'm wrong :) ):
```
const useGeneratedAnchor = ( clientId, anchor, content ) => {
1. inside here you would have `useSelect` to fetch `allHeadingAnchors` instead of doing it in `heading/edit.js`
2. execute the needed logic as is now, to return an `anchor`
}
```

So in `heading/edit.js` you would just import this hook and call it which would return the new `anchor`. Also there you would keep your `useEffect`, which would have as dependencies the `attributes.anchor` and the `generated one` and would `setAttributes` only if they are different.

Makes sense? I'm happy to help with anything Ari if you need to :)

Reply:
Thanks for the detailed instructions!
I believe I fixed it, could you please check now? :+1: